crates/typst-cli/src/compile.rs

```diff
@@ -227,7 +228,19 @@ pub fn print_diagnostics(
         config.display_style = term::DisplayStyle::Short;
     }
 
-    for diagnostic in warnings.iter().chain(errors) {
+    let mut unique_diags = HashSet::new();
+    let warnings: Vec<_> = warnings
+        .iter()
+        .filter(|item| unique_diags.insert((item.span, item.message.as_str())))
+        .collect();
+
+    unique_diags.clear();
+    let errors: Vec<_> = errors
+        .iter()
+        .filter(|item| unique_diags.insert((item.span, item.message.as_str())))
+        .collect();
```

Comment:
No need to collect into two vectors if they are chained anyway. Actually no need to collect at all since we can just check in the main printing loop if we've seen the same diagnostic before. See also `tracer.rs` which already deduplicates warnings. We could do the same here with a `HashSet<u128>` that keeps the full hashes of the diagnostics.

Reply:
See last comment.